crates/kornia-3d/src/voxelgrid.rs

```diff
@@ -0,0 +1,234 @@
+
+        for point in pointcloud.points() {
+            let point_vec = PointCloud::point_to_vec3(point);
+            if let Some(voxel_index) = self.get_voxel(&point_vec) {
+                voxel_points
+                    .entry(voxel_index)
+                    .or_insert_with(Vec::new)
+                    .push(point_vec);
+            }
+        }
+
+        let downsampled_points = voxel_points
+            .values()
+            .map(|points| {
+                let centroid = points.iter().fold(Vec3::ZERO, |acc, p| acc + *p) / points.len() as f32;
+                PointCloud::vec3_to_point(centroid)
+            })
+            .collect();
+
+        PointCloud::new(downsampled_points, None, None)
+    }
+
+    /// Create a VoxelGrid from a point cloud within specified bounds.
+    pub fn create_from_pointcloud_within_bounds(
+        pointcloud: &PointCloud,
+        voxel_size: f32,
+        min_bound: Vec3,
+        max_bound: Vec3,
+    ) -> Self {
+        if voxel_size <= 0.0 {
+            panic!("voxel_size must be greater than 0.");
+        }
+
+        if voxel_size * (i64::MAX as f32) < (max_bound - min_bound).max_element() {
+            panic!("voxel_size is too small.");
+        }
+
+        let mut voxel_grid = Self::new(voxel_size, min_bound);
+        let mut voxel_points: HashMap<[i64; 3], Vec<Vec3>> = HashMap::new();
+
+        for point in pointcloud.points() {
+            let point_vec = PointCloud::point_to_vec3(point);
+            if point_vec.cmpge(min_bound).all() && point_vec.cmple(max_bound).all() {
+                let relative_point = (point_vec - min_bound) / voxel_size;
+                let voxel_index = [
+                    relative_point.x.floor() as i64,
+                    relative_point.y.floor() as i64,
+                    relative_point.z.floor() as i64,
+                ];
+                voxel_points
+                    .entry(voxel_index)
+                    .or_insert_with(Vec::new)
+                    .push(point_vec);
+            }
+        }
+
+        for (voxel_index, points) in voxel_points {
+            let centroid = points.iter().fold(Vec3::ZERO, |acc, p| acc + *p) / points.len() as f32;
+            voxel_grid.voxels.insert(
+                voxel_index,
+                Voxel {
+                    grid_index: voxel_index,
+                    center: centroid,
+                },
+            );
+        }
+
+        voxel_grid
+    }
+
+    /// Create a VoxelGrid from a point cloud with automatic bounds.
+    pub fn create_from_pointcloud(
+        pointcloud: &PointCloud,
+        voxel_size: f32,
+    ) -> Self {
+        let min_bound = pointcloud.get_min_bound() - Vec3::splat(voxel_size * 0.5);
+        let max_bound = pointcloud.get_max_bound() + Vec3::splat(voxel_size * 0.5);
+        Self::create_from_pointcloud_within_bounds(pointcloud, voxel_size, min_bound, max_bound)
+    }
+}
+
+#[test]
+fn test_voxel_grid() {
```

Comment:
split this test in different; one for basic stuff (that you missed), check empty, min/max bound, add/remove, get_voxel, center coordinate, etc. Then another checking more modified geometry with downsample. Possibly a third testing the `create_from_pointcloud` and `create_from_pointcloud_within_bounds` functions